src/webgpu/api/operation/rendering/basic.spec.ts

```diff
@@ -129,4 +137,191 @@ g.test('large_draw')
       .combine('indexed', [true, false])
       .combine('indirect', [true, false])
   )
-  .unimplemented();
+  .fn(async t => {
+    const { indexed, indirect } = t.params;
+
+    const BYTES_PER_ROW = 256;
+    const dst = t.device.createBuffer({
+      size: 3 * BYTES_PER_ROW,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+
+    const indirectBuffer = t.device.createBuffer({
+      size: 20,
+      usage: GPUBufferUsage.INDIRECT | GPUBufferUsage.COPY_DST,
+    });
+    const writeIndirectParams = (count: number, instanceCount: number) => {
+      const params = new Uint32Array(5);
+      params[0] = count; // Vertex or index count
+      params[1] = instanceCount;
+      params[2] = 0; // First vertex or index
+      params[3] = 0; // First instance (non-indexed) or base vertex (indexed)
+      params[4] = 0; // First instance (indexed)
+      t.device.queue.writeBuffer(indirectBuffer, 0, params, 0, 5);
+    };
+
+    const MILLION = 1024 * 1024;
+    const MAX_INDICES = 16 * MILLION;
+    const indexBuffer = t.device.createBuffer({
+      size: MAX_INDICES * Uint32Array.BYTES_PER_ELEMENT,
+      usage: GPUBufferUsage.INDEX | GPUBufferUsage.COPY_DST,
+    });
+    const indexData = new Uint32Array(MILLION);
+    for (let offset = 0; offset < MAX_INDICES; offset += MILLION) {
+      for (let k = 0; k < MILLION; ++k) {
+        indexData[k] = offset + k;
+      }
+      t.device.queue.writeBuffer(
+        indexBuffer,
+        offset * Uint32Array.BYTES_PER_ELEMENT,
+        indexData,
+        0,
+        MILLION
+      );
+    }
+
+    const colorAttachment = t.device.createTexture({
+      format: 'rgba8unorm',
+      size: { width: 3, height: 3, depthOrArrayLayers: 1 },
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+    const colorAttachmentView = colorAttachment.createView();
+
+    const runPipeline = async (numInstances: number, numVertices: number) => {
+      const pipeline = t.device.createRenderPipeline({
+        vertex: {
+          module: t.device.createShaderModule({
+            code: `
+            struct Output {
+              [[builtin(position)]] position: vec4<f32>;
+              [[location(0)]] color: vec4<f32>;
+            };
+
+            fn selectValue(index: u32, maxIndex: u32) -> f32 {
+              let highOrMid = select(1.0, 0.5, index == maxIndex - 1u);
+              return select(0.0, highOrMid, index == 0u);
+            }
+
+            [[stage(vertex)]] fn main(
+                [[builtin(vertex_index)]] v: u32,
+                [[builtin(instance_index)]] i: u32)
+                -> Output {
+              let r = selectValue(v, ${numVertices}u);
+              let b = selectValue(i, ${numInstances}u);
+              let x = (r * 4.0 - 2.0) / 3.0;
+              let y = (b * 4.0 - 2.0) / -3.0;
+              return Output(vec4<f32>(x, y, 0.0, 1.0),
+                            vec4<f32>(r, 0.0, b, 1.0));
```

Comment:
optional: Hm, is it necessary to output different colors for the points? I think the test coverage would be roughly as good if the output color were constant - and it might make the shader execution slightly faster, enabling the test iterations to achieve higher draw call sizes before reaching the time limit.

Reply:
That's a fair point. If all fragments are covered, it's a reasonable signal that all vertices were processed.